### PR TITLE
[FIX] core: correctly order by nullable many2one fields 

### PR DESCRIPTION
**Pull request by aj-fuentes**
When ordering by a many2one field the ORM extends the ORDER BY clause
with the `_order_` field of the target model. This causes issues when
ordering via a many2one field if we modify the `_order` of the target
model **even** if the modification does not change the internal order of
the target model.

The reason is that when we order by a nullable many2one we may effective
add null values to order over. That can change the order. This is
especially relevant if DESC is used in the order of the target model.

For example let's say we have
```
+----------------------+       +---------------------+
|        Foo           |<--+   |        Bar          |
+----------------------+   |   +---------------------+
|_order = 'id'         |   +---|foo_id: many2one(Foo)|
|fool: Char -> required|       +---------------------+
+----------------------+
```

Then if we do `Bar.search([], order="foo_id")` we may get
inconsistent ordering. By consistent here we understand that if none of
the elements of `Foo` model changed its relative order then the elements
of `Bar` must also keep their relative order. Before this patch this was
not true as the added tests shows.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by aj-fuentes** [comment]
In principle nothing is wrong with ignoring null values order. But we recently faced an issue during an upgrade (see https://github.com/odoo/upgrade/pull/3791)

The situation is that if we change the _order slot of a model A (from version X to X+1) even if we are sure that for A the ordering is kept across upgrades we may introduce order inconsistencies for models that order by `a_id` (via many2one). An example of such change is odoo/odoo@687c2060 which makes [14.0](https://github.com/odoo/odoo/blob/657fd6653705255f319756cba6406cefed9649d4/addons/mrp/models/mrp_bom.py#L221) return a different BoM than in [15.0](https://github.com/odoo/odoo/blob/7877abd7fc6bc42f4a6bbff28d840135b12e9885/addons/mrp/models/mrp_bom.py#L249) (this then leads to different quantities being reported for the same product post-upgrade).

I will work on trying to reproduce this real case in the test...

**Comment by ryv-odoo** [comment]
Hello @aj-fuentes , Indeed there is something not perfectly right about _order.

There are already some issues when the `_order` doesn't represent a unique way to order records without ambiguity (the order key should be never equals).
When a model doesn't have deterministic `_order`, two same `search` execution can return two different result order (because PostgreSQL doesn't ensure that). In general (always?), we want to avoid this nondeterministic behavior. (@Xavier-Do )

Your case described is a sub-case of the issue explained here and you shouldn't fix it like this (we can have multiple time id at the order end (noise), it doesn't cover the simple case, ...).

We should maybe enforce (or correct it for each model) that `_order` is always deterministic. What do you think ?

**Comment by aj-fuentes** [comment]
Hello @ryv-odoo 
> Your case described is a sub-case of the issue explained here and you shouldn't fix it like this

The goal of this (draft) PR is to highlight the issue first, with a _probable_ implementation :). Thus as I described in the commit message: we can try to improve the order for many2ones (it won't fix everything, just improve, if any, many2one ordering).

My point I guess is: when we order by `model_id` (many2one) I agree that we must add the `_order` specs of the target model, but it's also natural to expect that in case of ties we do not treat both categories of null values the same. Moreover, with all the wreckage this may cause, I would say that we should always implicitly include the `id` of a record after we expand the `_order` spec into ORDER BY. That way we have a sure tie-breaker, that will matter only for such models where the columns in `_order` are already non-unique. A special case of this is PR (thus why I believe it highlights more nuances of this issue). In the end I proposed the extra id for many2ones only, but I think we may consider instead adding `id` always.

> We should maybe enforce (or correct it for each model) that _order is always deterministic. What do you think ?

I agree, hence the implicit `id` proposal.

> it doesn't cover the simple case

I'm not sure what simple case you meant, maybe it's already on what I wrote above :)

A final word on why I believe the many2one issue is more delicate. The reason is that the many2one `_order` expansion is transparent for the users[^1]. This means that it's harder for developers to understand what may happen when they add a column to `_order`. If they are diligent they may double check that the issue keeps a good enough order for the current model. Still it's hard if not impossible to think about other models ordering by a many2one to it. That's because there is a new category of nulls (when the FK is allowed to be null) being added, for which the developer never had a chance to think about. Such category will be intermixed with the actual null values and thus breaking the original assumptions of the developer. Therefore it may also break the default PG behavior one may rely on: nulls are ordered last for ASC (but in a contiguous block) or the first for DESC (contiguous block). The null FK introduces null values that may inject themselves within the contiguous block of original null values.

(may views on this topic my be wrong anyway)

[^1]: I had a hard time debugging why the order was different in the linked upgrade.

**Comment by aj-fuentes** [comment]
For the sake of the discussion this is the query that this changes outputs:
```sql
SELECT "test_new_api_bar_order".id FROM "test_new_api_bar_order" LEFT JOIN
"test_new_api_foo_order" AS "test_new_api_bar_order__foo_id" ON
("test_new_api_bar_order"."foo_id" = "test_new_api_bar_order__foo_id"."id") WHERE
("test_new_api_bar_order"."id" in (25, 26)) ORDER BY  
"test_new_api_bar_order__foo_id"."foo" ,"test_new_api_bar_order__foo_id".id
```
vs (without)
```sql
SELECT "test_new_api_bar_order".id FROM "test_new_api_bar_order" LEFT JOIN
"test_new_api_foo_order" AS "test_new_api_bar_order__foo_id" ON 
("test_new_api_bar_order"."foo_id" = "test_new_api_bar_order__foo_id"."id") WHERE 
("test_new_api_bar_order"."id" in (27, 28)) ORDER BY  
"test_new_api_bar_order__foo_id"."foo"
```

**Comment by ryv-odoo** [comment]
Sorry my previous comment is unclear. 

> The goal of this (draft) PR is to highlight the issue first

Thank you about that :)

> it's also natural to expect that in case of ties we do not treat both categories of null values the same

Totally agree

> I'm not sure what simple case you meant, maybe it's already on what I wrote above :)

By "Simple case" (=> related case) I mean: if the `_order` of a model doesn't threat tie correctly (not unique order),
the result of single search on this model is already nondeterministic. IMO, it is the bigger issue about `_order`.
In your example, the model `test_new_api_bar_order` has a "bad" `_order` because it doesn't threat ties correctly and should be set at 'foo, id' (which by extension solve your issue transparently, or I miss something ?). 

> A final word on why I believe the many2one issue is more delicate. The reason is that the many2one `_order` expansion is transparent for the users. 

Yes, Indeed lot of people are not aware of that.


My goal was to propose a solution (Force correct and unambiguous `_order`) which cover your case but also the case of nondeterministic result from a `search` (which can lead to nondeterministic behavior functionally).

**Comment by aj-fuentes** [comment]
OK upon carefully re-thinking I was able to distill what I actually wanted to. I've updated the test and the fix. Thanks @KangOl for pointing out how to patch the _order of a record.

TL;DR: if we change _order in a model A in such a way that the order of the records is NOT changed, we may still change the order of a model B when ordering via a many2one to A. I proposed a solution to discuss.

**Comment by C3POdoo** [comment]
Dear @aj-fuentes,

Thank you for your contribution but we are closing it due to inactivity.
We only support the last 3 stable versions so no longer accepts patches into this branch.

We apology if we could not look at your request in time.
If the contribution still makes sense, please let us know and do not hesitate to recreate one for the recent versions. We will try to check it as soon as possible.

_This is an automated message._
